README.md

```diff
@@ -112,52 +136,111 @@ func main() {
+options offered by the SDK because it allows seamless use of both service
+principals and [Azure Managed Service Identity][]. Other options are listed
+below.
+
+> Note: If you need to create a new service principal, run `az ad sp
+> create-for-rbac -n "<app_name>"` in the
+> [azure-cli](https://github.com/Azure/azure-cli). See [these
+> docs](https://docs.microsoft.com/cli/azure/create-an-azure-service-principal-azure-cli?view=azure-cli-latest)
+> for more info. Copy the new principal's ID, secret, and tenant ID for use in
+> your app, or consider the `--sdk-auth` parameter for serialized output.
+
+[Azure Managed Service Identity]: https://docs.microsoft.com/en-us/azure/active-directory/msi-overview
+
+* The `auth.NewAuthorizerFromEnvironment()` described above creates an authorizer
+from the first available of the following configuration:
+
+    1. **Client Credentials**: Azure AD Application ID and Secret.
+
+        - `AZURE_TENANT_ID`: Specifies the Tenant to which to authenticate.
+        - `AZURE_CLIENT_ID`: Specifies the app client ID to use.
+        - `AZURE_CLIENT_SECRET`: Specifies the app secret to use.
+
+    2. **Client Certificate**: Azure AD Application ID and X.509 Certificate.
+
+        - `AZURE_TENANT_ID`: Specifies the Tenant to which to authenticate.
+        - `AZURE_CLIENT_ID`: Specifies the app client ID to use.
+        - `AZURE_CERTIFICATE_PATH`: Specifies the certificate Path to use.
+        - `AZURE_CERTIFICATE_PASSWORD`: Specifies the certificate password to use.
+
+    3. **Resource Owner**: Azure AD User and Password. This grant type is *not
+       recommended*, use client credentials (service principals) instead.
+
+        - `AZURE_TENANT_ID`: Specifies the Tenant to which to authenticate.
+        - `AZURE_CLIENT_ID`: Specifies the app client ID to use.
+        - `AZURE_USERNAME`: Specifies the username to use.
+        - `AZURE_PASSWORD`: Specifies the password to use.
+
+    4. **Azure Managed Service Identity**: Delegate credential management to the
+       platform. Requires that code is running in Azure, e.g. on a VM. All
+       configuration is handled by Azure. See [Azure Managed Service
+       Identity](https://docs.microsoft.com/en-us/azure/active-directory/msi-overview)
+       for more details.
+
+* The `auth.NewAuthorizerFromFile()` method creates an authorizer using tokens
+  serialized and cached by the [Azure CLI][]. Follow these steps to utilize:
+
+    1. Create a service principal with serialized output file using `az ad sp
+       create-for-rbac --sdk-auth > client_credentials.json`.
+    2. Set environment variable `AZURE_AUTH_LOCATION` to the path of the saved
+       output file.
+    3. Use the authorizer returned by `auth.NewAuthorizerFromFile()` in your
+       client as described above.
+
+[Azure CLI]: https://github.com/Azure/azure-cli
+
+* Finally, you can use OAuth's [Device Flow][] by calling
+  `auth.NewDeviceFlowConfig()` and extracting the Authorizer as follows:
```

Comment:
>NewDeviceFlowConfig [](start = 8, length = 19)

maybe specify that this is for 2FA access for people who are not aware.

Reply:
@vladbarosan 2FA is dependent on the user account that's logging in, not the login method utilized. For example, if I sign in with authz code flow I also have to approve via 2FA if it's configured for my user account. If I don't have 2FA enabled I don't need it for device flow either.

Reply:
👍